Improve Jupyter notebook converter's handling of code blocks (#940)
sphinx-gallery/sphinx-gallery: doc/configuration.rst

See the Sphinx-Gallery `Sphinx configuration file <https://github.com/sphinx-gallery/sphinx-gallery/blob/master/doc/conf.py>`_
for an example that uses the `public Binder server <http://mybinder.org>`_.

.. _promote_jupyter_magic:

> Make Jupyter cell magic compile as code

Author: How about "Making cell magic executable in notebooks"? Most of the other titles have the verb as a present participle (e.g. `Compressing images`, `Building without executing examples`, `Choosing the thumbnail image`).

See the Sphinx-Gallery `Sphinx configuration file <https://github.com/sphinx-gallery/sphinx-gallery/blob/master/doc/conf.py>`_
for an example that uses the `public Binder server <http://mybinder.org>`_.

.. _promote_jupyter_magic:

Making cell magic executable in notebooks
=========================================

> Often times, tutorials will include bash scripts for the user to copy/paste into

Reviewer: ```suggestion
Often times, tutorials will include bash code for the user to copy/paste into
```
not sure, script just has connotations of file to me.

their terminal. These scripts are not intended to be run when someone is
building the documentation, as they will already have those dependencies in
their environment. Hence they are normally written as code blocks inside text::

  #%%
  # Installing dependencies
  #
  #     .. code-block:: bash
  #
  #       pip install -q tensorflow
  #       apt-get -qq install curl

This works fine for the ``.py`` and ``.html`` files, but causes problems when
rendered as an Jupyter notebook. The downloaded ``.ipynb`` file will not have

> those dependencies installed, and thus will not work without running the script.

Reviewer: ```suggestion
those dependencies installed, and thus will not work without running the bash code.
```

To fix this, we can set the ``promote_jupyter_magic`` flag in ``conf.py``::

  sphinx_gallery_conf = {
      ...
      'promote_jupyter_magic': True,
  }

If this flag is ``True``, then when a Jupyter notebook is being built, any code

> block starting with a `Jupyter cell magic <https://ipython.readthedocs.io/en/stable/interactive/magics.html>`_ (e.g. ``%%bash`` or ``%%writefile``)

Reviewer: ```suggestion
block starting with `Jupyter cell magic <https://ipython.readthedocs.io/en/stable/interactive/magics.html>`_ (e.g. ``%%bash`` or ``%%writefile``)
```

Author: Changed to "blocks starting with Jupyter cell magics", as [the IPython docs use "magics" as the plural](https://ipython.readthedocs.io/en/stable/interactive/magics.html).

will be turned into a runnable code block.

For our earlier example, we could change the Markdown text to::

  #%%
  # Installing dependencies
  #
  #     .. code-block:: bash
  #
  #       %%bash
  #       pip install -q tensorflow
  #       apt-get -qq install curl

> Then, TensorFlow and Curl would be automatically installed upon running the

Reviewer: ```suggestion
meaning, TensorFlow and Curl would be automatically installed upon running the
```

Jupyter notebook. This works for any cell magic (not just those mentioned above)
and only affects the creation of Jupyter notebooks.

.. warning::
  It is good practice to ensure the ``.py`` and ``.html`` files match the ``.ipynb``

>   files as closely as possible. This flag (and Jupyter cell magics) should only

Reviewer: ```suggestion
  files as closely as possible. This functionality should only
```
